webrtc.html

```diff
@@ -509,33 +509,43 @@
 
         <ol>
           <li>
-            <p>Append an object representing the current call being handled
-            (i.e. function name and corresponding arguments) to the
-            <code>operations</code> array.</p>
+            <p>If <var>connection</var>'s <a href=
+              "#dom-peerconnection-signaling-state"><code>RTCPeerConnection</code>
+              signalingState</a> is <code>closed</code>, throw an
+              <code>InvalidStateError</code> exception and abort these steps.</p>
+          </li>
+
+          <li>
+            <p>Let <var>p</var> be the promise that the called
+            function will return.
+            </p>
+          </li>
+            <p>Add <code>p</code> to <var>operations</var>.</p>
           </li>
 
           <li>
-            <p>If the length of the <code>operations</code> array is exactly 1,
-            execute the function from the front of the queue
-            asynchronously.</p>
+            <p>If the length of <var>operations</var> is exactly 1,
+            execute the steps necessary to resolve <code>p</code>.</p>
           </li>
 
           <li>
-            <p>When the asynchronous operation completes (either successfully
-            or with an error), remove the corresponding object from the
-            <code>operations</code> array. After removal, if the array is
-            non-empty, execute the first object queued asynchronously and
-            repeat this step on completion.</p>
+            <p>When a promise in <code>operations</code> is resolved,
+            remove it from <code>operations</code>.  After removal,
+            if <code>operations</code> is non-empty, execute the steps
+            necessary to resolve the first promise
+            in <code>operations</code>.
```

Comment:
The argument that JavaScript developers benefit from simple implementations, I think is a red herring here. JavaScript Developers benefit from cross-browser deterministic behavior.

Simplicity for implementers is low in the [priority of constituencies](http://www.w3.org/TR/html-design-principles/#priority-of-constituencies).

> signaling.onoffer = offer => pc.setRemoteDescription(offer).then(=> pc.createAnswer(offer). then(answer => pc.setLocalDescription(answer)));

Btw, I think your example is non-deterministic, because it seems like a genuine race between `signaling.onoffer` and createOffer finishing.

Still, I think once one realizes that every `p.then(func)` indirection amounts to another (guaranteed) spin of the microtask queue, then one quickly becomes aware of timing issues like "what order will a and b run in given `p1.then(() => {}).then(a)` and `p1.then(b)`.

The answer may not always need to be obvious, but it better be deterministic, and hopefully across browsers, or debugging just got a lot harder.